File: cpp/src/solutions/valid_anagram.cpp
// LeetCode 242. Valid Anagram in C++

#include "cpp/src/common/common.h"

class Solution {
 public:
  bool isAnagram(string s, string t) {
    vector<int> flag(26, 0);
    for_each(s.begin(), s.end(), [&](char &ch) { flag[ch - 'a'] += 1; });
    for_each(t.begin(), t.end(), [&](char &ch) { flag[ch - 'a'] -= 1; });
    auto valid = count_if(flag.begin(), flag.end(),
                          [](int &value) { return value != 0; });
    return valid == 0;
  }
};

TEST(isAnagram, example1) {
  string s = "anagram", t = "nagaram";
  bool ans = true;
  EXPECT_EQ(Solution().isAnagram(s, t), ans);
}

TEST(isAnagram, example2) {
  string s = "rat", t = "car";
  bool ans = false;
  EXPECT_EQ(Solution().isAnagram(s, t), ans);
}

File: rust/src/solutions/valid_anagram.rs
// LeetCode 242. Valid Anagram

use super::Solution;

impl Solution {
    pub fn is_anagram(s: String, t: String) -> bool {
        let mut flag = vec![0; 26];
        for ch in s.bytes() {
            flag[ch as usize - 'a' as usize] += 1;
        }
        for ch in t.bytes() {
            flag[ch as usize - 'a' as usize] -= 1;
        }
        flag.iter().filter(|&x| *x != 0).count() == 0
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn example1() {
        let s = String::from("anagram");
        let t = String::from("nagaram");
        let ans = true;
        assert_eq!(Solution::is_anagram(s, t), ans);
    }

    #[test]
    fn example2() {
        let s = String::from("rat");
        let t = String::from("car");
        let ans = false;
        assert_eq!(Solution::is_anagram(s, t), ans);
    }
}
